theano/gpuarray/basic_ops.py

```diff
@@ -313,6 +320,90 @@ def kernel_version(self, node):
         return (4, self.get_params(node).bin_id)
 
 
+def fwds(name):
+    def f(*args):
+        res = getattr(GpuKernelBase, name)(*args)
+        try:
+            res = res + '\n' + getattr(COp, name)(*args)
+        except MethodNotDefined:
+            pass
+        return res
+    f.__name__ = name
+    return f
+
+
+class CGpuKernelBase(COp, GpuKernelBase):
+    """
+    Class to combine GpuKernelBase and COp.
+
+    It adds a new section type 'kernels' where you can define kernels
+    with the '#kernel' tag
+    """
+    SECTIONS = copy.copy(COp.SECTIONS)
+    SECTIONS.add('kernels')
+
+    kernel_re = re.compile(r'^#kernel ([a-zA-Z_].*?)$', re.MULTILINE)
+
+    c_support_code = fwds('c_support_code')
+    c_support_code_apply = fwds('c_support_code_apply')
+    c_support_code_struct = fwds('c_support_code_struct')
+    c_init_code_struct = fwds('c_init_code_struct')
+    c_cleanup_code_struct = fwds('c_cleanup_code_struct')
+
+    def _type_macros(self, node):
```

Comment:
I don't like that. This push in the old direction on compiling code for different dtypes. Do you have an idea how to allow the same thunk to support multiple-dtype? Ideally, we should have the current user interface support a possible futur extension that will compile less thunk.